src/Controller/Admin/Improve/Design/LinkWidgetController.php

```diff
@@ -0,0 +1,252 @@
+     */
+    public function editAction($linkBlockId)
+    {
+        $this->get('prestashop.module.link_block.form_provider')->setIdLinkBlock($linkBlockId);
+        $form = $this->get('prestashop.module.link_block.form_handler')->getForm();
+
+        return [
+            'linkBlockForm' => $form->createView(),
+        ];
+    }
+
+    /**
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="Access denied.")
+     * @Template("@Modules/ps_linklist/views/templates/admin/link_block/form.html.twig")
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse|array
+     * @throws \Exception
+     */
+    public function createProcessAction(Request $request)
+    {
+        return $this->processForm($request, 'Successful creation.');
+    }
+
+    /**
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="Access denied.")
+     * @Template("@Modules/ps_linklist/views/templates/admin/link_block/form.html.twig")
+     *
+     * @param Request $request
+     * @param $linkBlockId
+     *
+     * @return RedirectResponse|array
+     * @throws \Exception
+     */
+    public function editProcessAction(Request $request, $linkBlockId)
+    {
+        return $this->processForm($request, 'Successful update.', $linkBlockId);
+    }
+
+    /**
+     * @param int $linkBlockId
+     * @return RedirectResponse
+     */
+    public function deleteAction($linkBlockId)
+    {
+        $repository = $this->get('prestashop.module.link_block.repository');
+        $errors = [];
+        try {
+            $repository->delete($linkBlockId);
+        } catch (PrestaShopDatabaseException $e) {
+            $errors[] = [
+                'key' => "Could not delete #%i",
+                'domain' => 'Admin.Catalog.Notification',
+                'parameters' => [$linkBlockId],
+            ];
+        }
+
+        if (0 === count($errors)) {
+            $this->clearModuleCache();
+            $this->addFlash('success', $this->trans('Successful deletion.', 'Admin.Notifications.Success'));
+        } else {
+            $this->flashErrors($errors);
+        }
+
+        return $this->redirectToRoute('admin_link_widget_list');
+    }
+
+    /**
+     * @param Request  $request
+     * @param string   $successMessage
+     * @param int|null $linkBlockId
+     * @return array|RedirectResponse
+     * @throws \Exception
+     */
+    private function processForm(Request $request, $successMessage, $linkBlockId = null)
+    {
+        /** @var LinkBlockFormDataProvider $formProvider */
+        $formProvider = $this->get('prestashop.module.link_block.form_provider');
+        $formProvider->setIdLinkBlock($linkBlockId);
+
+        /** @var FormHandlerInterface $formHandler */
+        $formHandler = $this->get('prestashop.module.link_block.form_handler');
+        $form = $formHandler->getForm();
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted()) {
```

Comment:
Missing `$form->isValid();` ?

Reply:
yes that's why I added the isValid, just in case
Even though in this case the validation is performed by the FormHandler in `saveData` method, not a big fan of this though maybe the Interface should implement a `validateData` method